### ISSUE TITLE
GenerateCostDetails does not delete inactive cost types or methods

### ISSUE THREAD

**Issue by mckayERP** [adempiere/adempiere#197]
For a cost type, create two methods for material costs.  Create some MR or internal use inventory records, note that there are duplicate postings of the material costs - one for each method.  Now try to get rid of the extra material cost and return to a single posting per transaction.  I can`t figure out how to do this with the current software.

I`ve debugged and found that MCostDetails generates costs for all methods - including inactive ones while the GenerateCostDetails only deletes and recreates the active ones.  There is no way to delete cost details for inactive cost methods.

**Comment by e-Evolution** [comment]
Hi Mike

Yes now the cost detail is generate for each cost type , for this reason that now Adempiere support costing in parallel , it is very useful when we need know the cost from different respective , but the accounting the system take the cost type defined in account schema to create posting facts 

I reviewed the code and currently the system only generate for active cost type 

https://github.com/adempiere/adempiere/blob/hotfix/3.8.0%23002/base/src/org/compiere/model/MCostType.java#L90

So I not understand how you get this behavior.

have you some code line to review :-)

**Comment by mckayERP** [comment]
The problem is it only deletes active cost type and methods as well.  So you can't delete the cost details for the types and methods that are inactive.  The loop to delete the cost details needs to include the inactive records as well.

**Comment by mckayERP** [comment]
Here are more detailed steps to generate the error.

Log in to GardenWorld with GardenWorld Admin role.
Generate cost details and run the client accounting processor.
Find the cost for the plum tree.  Note average invoice =  $41.19 and Default = $20.
Open the Cost Element window.
Find the Cost Element with Name = "Average Invoice" and Cost Element Type =  "Material"- it should be inactive. Make it active.
Regenerate the cost details for all cost types, elements and the plum tree.
Check the Product Cost Dimensions.  Note that "Average Invoice" appears as a cost element in both the Default and Average Invoice cost types.
Now try to delete these entries.
Find the record made active and try to delete it.  Not allowed.  Make it inactive again.
Regenerate the cost details for all cost types.
Check the Product Cost Dimensions and note that the inactive Cost Element is still listed.
Now I'm stuck.

**Comment by mckayERP** [comment]
To fix it, I've added several lists to GenerateCostDetail that will hold the cost types and elements that will be deleted.  These are potentially different from the ones that will be created as they will include the inactive types and elements.  The delete loop uses these new lists to delete the cost details.

I've modified the MCostElement.beforeSave to ensure there is only one active record of each cost type per client and only one default element per client.

I've also modified the MCostElement.beforeDelete to delete both the related cost details and cost dimensions.  This makes it very easy to correct problems when setting up.  I couldn't think of a reason not to allow this.  The cost details and dimensions are not part of the audit trail and the details are deleted and regenerated as required.

There was some left over code in the MCostElement related to cost elements defining costing methods.  I deprecated this as well and removed references to it in MCost in 1 place.  I think the cost method column can be deleted from the M_CostElement table in a future release.

I'd also propose deleting the wrong entries in the GW client so as to not confuse new users.

**Comment by e-Evolution** [comment]
Hi Mike @mckayERP
Check the code quickly and I see that the logic of only allowing a record in the cost dimension by type of cost element has been added

https://github.com/adempiere/adempiere/pull/2384/files#diff-0000b87d8d78368608a5f798e4fe656fR240

Which is not correct, that is, the cost system is designed so that you can add multiple combinations in the combination of the cost dimension, even several elements on the same type of cost element.

Example of real life.

It is common that in the marketers import products from abroad and want to have a detailed landed cost type analysis, example I should have the possibility, in addition, to adding three elements of cost with the landed cost element cost type, to have the cost of freight, the costs of import and indirect by manage with same type of cost element but with cost element separately, in this case, Landed cost.

Thus the logic of having only one item of cost per type is not correct.

Please fix or revert this pull request

kind regards

**Comment by yamelsenih** [comment]
Hi, any update for it?
